contracts/staking/test/unit_tests/lib_proxy.ts

```diff
@@ -0,0 +1,324 @@
+
+            it('should revert when the destination is address zero', async () => {
+                checkDestinationZeroError(await publicProxyCallAsync({ destination: constants.NULL_ADDRESS }));
+            });
+
+            it('should revert when the destination is address zero and revertRule == AlwaysRevert', async () => {
+                checkDestinationZeroError(
+                    await publicProxyCallAsync({
+                        destination: constants.NULL_ADDRESS,
+                        revertRule: RevertRule.AlwaysRevert,
+                    }),
+                );
+            });
+
+            it('should revert when the destination is address zero and revertRule == NeverRevert', async () => {
+                checkDestinationZeroError(
+                    await publicProxyCallAsync({
+                        destination: constants.NULL_ADDRESS,
+                        revertRule: RevertRule.NeverRevert,
+                    }),
+                );
+            });
+        });
+
+        describe('Calldata Checks', () => {
+            it('should simply forward the calldata and succeed when customEngressSelector == bytes4(0), ignoreIngressSelector == false, and revertRule = RevertOnError', async () => {
+                const calldata = constructRandomSuccessCalldata();
+
+                // Ensure that the returndata (the provided calldata) is correct.
+                checkEndingConditions(await publicProxyCallAsync({ calldata }), true, calldata);
+            });
+
+            it('should send the customEgressSelector followed by the calldata when customEgressSelector != bytes4(0), ignoreIngressSelector == false, and revertRule == RevertOnError', async () => {
+                const calldata = constructRandomSuccessCalldata();
+
+                // Choose a random customEgressSelector selector.
+                const customEgressSelector = randomBytes4();
+
+                // Ensure that the returndata (the provided calldata) is correct.
+                checkEndingConditions(
+                    await publicProxyCallAsync({
+                        calldata,
+                        customEgressSelector,
+                    }),
+                    true,
+                    customEgressSelector.concat(calldata.slice(2, calldata.length)),
+                );
+            });
+
+            it('should send the the calldata without the selector when customEgressSelector == bytes4(0), ignoreIngressSelector == true, and revertRule == RevertOnError', async () => {
+                const calldata = constructRandomSuccessCalldata();
+
+                // Ensure that the returndata (the provided calldata) is correct.
+                checkEndingConditions(
+                    await publicProxyCallAsync({
+                        calldata,
+                        ignoreIngressSelector: true,
+                    }),
+                    true,
+                    '0x'.concat(calldata.slice(10, calldata.length)),
```

Comment:
We should add a `hexSlice()` function to `hex_utils`, so we can just do `hexSlice(calldata, 4)`.

Reply:
Yeah I'll go ahead and do that. It would be much nicer.